src/rust/engine/BUILD

```
@@ -0,0 +1,37 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+files(
+    name="rust_sources",
+    sources=[
+        "Cargo.lock",
+        "build.rs",
+        "**/Cargo.toml",
+        "**/*.rs",
+        "!**/*tests.rs",
+        "**/*.proto",
+        ".cargo/config",
+        "rust-toolchain",
+    ],
+)
+
+shell_sources(name="shell")
+
+shell_command(
+    name="engine-and-client",
+    command=f"./cargo_build_shim.sh --features=extension-module -p engine -p client",
+    execution_dependencies=[
+        ":rust_sources",
+        "./cargo_build_shim.sh:shell",
+        "3rdparty/tools/python3:python3",
+    ],
+    extra_env_vars=["CHROOT={chroot}", "MODE"],
+    tools=["cc", "ld", "as", "ar", "cargo", "protoc", "python3.9", "sccache"],
+    output_files=[
+        f"target/debug/libengine.so",
+        f"target/debug/pants",
+        f"target/release/libengine.so",
+        f"target/release/pants",
+    ],
```

I tried this locally and it seems to work pretty well (nice job!), except `dist/codegen` only contains `src/rust/engine/target/debug/pants`, not a `libengine.so`. Is the engine filename wrong here?

(https://github.com/pantsbuild/pants/issues/18183 is related.)

---

Ah, on Mac, the file is `libengine.dylib`, no `.so`.

```suggestion
    output_files=[
        f"target/debug/libengine.so",
        f"target/debug/libengine.dylib",
        f"target/debug/pants",
        f"target/release/libengine.so",
        f"target/release/libengine.dylib",
        f"target/release/pants",
    ],
```